### PR TITLE
Add API to fetch conflicting task locks

### PR DESCRIPTION
**Pull request by AmatyaAvadhanula**
### Description

Add a new API to view the active task locks for a given set of (datasource, priority, interval list).
This can be useful to:
1. View lock contention
2. Observe existence of concurrent locks
3. Augment and replace the existing /lockedIntervals API 
4. Miscellaneous debugging

### Changes
1. Remove the existing deprecated API to view locked intervals: `/lockedIntervals`
2. Augment lock filter policy to take a list of intervals
3. Add an API to view task locks given a set of filters instead: `/activeLocks`
4. Add tests and modify existing ones accordingly

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/http/TaskLockResponse.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  @JsonCreator
+  public TaskLockResponse(
+      @JsonProperty("taskLocks") final Map<String, List<TaskLockInfo>> taskLocks
+  )
+  {
+    this.taskLocks = taskLocks;
+  }
+
+  @JsonProperty
+  public Map<String, List<TaskLockInfo>> getTaskLocks()
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/http/TaskLockResponse.java:43]: Nit: either call it `datasourceToTaskLocks` (or similar) or add a javadoc.

File: server/src/main/java/org/apache/druid/rpc/indexing/OverlordClient.java [hunk 2]
```diff
@@ -180,15 +181,13 @@ ListenableFuture<CloseableIterator<TaskStatusPlus>> taskStatuses(
   ListenableFuture<CloseableIterator<SupervisorStatus>> supervisorStatuses();
 
   /**
-   * Returns a list of intervals locked by higher priority conflicting lock types
+   * Returns a list of Locks of higher priority with conflicting intervals
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/rpc/indexing/OverlordClient.java:184]:
```suggestion
   * Returns a list of locks of higher priority with conflicting intervals
```

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueryTool.java [hunk 3]
```diff
@@ -85,28 +86,12 @@ public TaskQueryTool(
   }
 
   /**
-   * @param lockFilterPolicies Requests for conflicing lock intervals for various datasources
-   * @return Map from datasource to intervals locked by tasks that have a conflicting lock type that cannot be revoked
+   * @param lockFilterPolicies Requests for conflicing locks for various datasources
+   * @return Map from datasource to conflicting lock infos
    */
-  public Map<String, List<Interval>> getLockedIntervals(List<LockFilterPolicy> lockFilterPolicies)
+  public Map<String, List<TaskLockInfo>> getConflictingLockInfos(List<LockFilterPolicy> lockFilterPolicies)
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueryTool.java:92]: Rename all the methods named `getConflictingLockInfos()` to `getActiveTaskLocks()`.
These APIs are just returning active locks which satisfy certain conditions.

The fact whether they conflict with something else or not is the caller's concern.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java [hunk 4]
```diff
@@ -241,33 +242,18 @@ public Response isLeader()
...
-      return Response.status(Status.BAD_REQUEST).entity("No Datasource provided").build();
-    }
-
-    // Build the response
-    return Response.ok(taskQueryTool.getLockedIntervals(minTaskPriority)).build();
-  }
-
   @POST
-  @Path("/lockedIntervals/v2")
+  @Path("/conflictingLocks")
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:246]: Call the new API `/activeLocks`.

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java [hunk 5]
```diff
@@ -241,33 +242,18 @@ public Response isLeader()
...
-    if (minTaskPriority == null || minTaskPriority.isEmpty()) {
-      return Response.status(Status.BAD_REQUEST).entity("No Datasource provided").build();
-    }
-
-    // Build the response
-    return Response.ok(taskQueryTool.getLockedIntervals(minTaskPriority)).build();
-  }
-
   @POST
-  @Path("/lockedIntervals/v2")
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:260]: Let's retain this API as it is not deprecated yet. We can mark it deprecated as a part of the changes in this PR.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/TaskLock.java [hunk 6]
```diff
@@ -77,4 +78,9 @@ default void assertNotRevoked()
           .build("Lock of type[%s] for interval[%s] was revoked", getType(), getInterval());
     }
   }
+
+  default TaskLockInfo toTaskLockInfo()
```
[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/TaskLock.java:82]: This is not possible as the class TaskLock is not available there.

[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/TaskLock.java:82]: Will add it as a static method in TaskLocks instead as you suggested offline

File: server/src/main/java/org/apache/druid/metadata/TaskLockInfo.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+{
+  private final String granularity;
+  private final String type;
+  private final int priority;
+  private final Interval interval;
+
+  @JsonCreator
+  public TaskLockInfo(
+      @JsonProperty("granularity") String granularity,
+      @JsonProperty("type") String type,
```
[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/metadata/TaskLockInfo.java:41]: The actual types aren't available in this module

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/metadata/TaskLockInfo.java:41]: Yeah, you can use lock type as String.
Granularity should still be available though, right?

[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/metadata/TaskLockInfo.java:41]: Granularity is LockGranularity here, which is unavailable

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/metadata/TaskLockInfo.java:41]: I see, thanks for the clarification.

If we are going to continue using the old API for the coordinator, does `TaskLockInfo` still need to be in `server` module?

[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/metadata/TaskLockInfo.java]: Can we just use the TaskLock object in that case?

[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/metadata/TaskLockInfo.java]: Yeah, that works too. 👍🏻

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java (hunk 4 above)
Comment (AmatyaAvadhanula) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java]: Done

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java (hunk 5 above)
Comment (AmatyaAvadhanula) [reply indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:260]: Done

File: server/src/main/java/org/apache/druid/server/http/TaskLockResponse.java (hunk 1 above)
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/server/http/TaskLockResponse.java]: Done

**Pull request merged**